dali/pipeline/operators/util/reshape.cc

```diff
@@ -0,0 +1,198 @@
+  bool has_layout_arg = spec.HasArgument("layout");
+  DALI_ENFORCE(!(has_shape_input && has_shape_arg),
+    "Reshape: use either shape input or shape argument, not both");
+  DALI_ENFORCE(has_shape_input || has_shape_arg || has_layout_arg,
+    "Reshape is no-op: arguments specify neither new shape nor layout.");
+  if (has_shape_arg) {
+    if (spec.HasTensorArgument("shape")) {
+      shape_source_ = ShapeSource::ArgInput;
+    } else {
+      auto shape_vec = spec.GetRepeatedArgument<int>("shape");
+      DALI_ENFORCE(!shape_vec.empty(), "Reshape: `shape` specified as empty list");
+      uniform_shape_.resize(shape_vec.size());
+      for (int i = 0; i < uniform_shape_.sample_dim(); i++) {
+        DALI_ENFORCE(shape_vec[i] > 0, "Reshape: all extents must be positive; got: " +
+            std::to_string(uniform_shape_));
+        uniform_shape_[i] = shape_vec[i];
+      }
+      shape_source_ = ShapeSource::Arg;
+    }
+  }
+  if (has_shape_input) {
+    DALI_ENFORCE(spec.InputDevice(1) == "cpu", "Output shapes must be provided as a CPU input");
+    shape_source_ = ShapeSource::Input;
+  }
+  if (has_layout_arg) {
+    layout_ = spec.GetArgument<TensorLayout>("layout");
+  }
+}
+
+template <typename Backend>
+bool Reshape<Backend>::SetupImpl(std::vector<OutputDesc> &output_desc, const Workspace &ws) {
+  output_desc.resize(1);
+  CalculateOutputShape(ws);
+  output_desc[0].type = ws.template InputRef<Backend>(0).type();
+  output_desc[0].shape = output_shape_;
+  return false;
+}
+
+template <typename Backend>
+template <typename Integer>
+void Reshape<Backend>::ShapeFromInput(
+      const kernels::TensorListView<kernels::StorageCPU, Integer> &shape) {
+  DALI_ENFORCE(shape.sample_dim() == 1 || (shape.sample_dim() == 2 && shape.num_samples() == 1),
+    "Reshape: shape input must be a list of 1D tensors or a single 2D tensor");
+  if (shape.sample_dim() == 2) {
+    auto shape_tensor = shape[0];
+    int N = shape_tensor.shape[0];
+    int dim = shape_tensor.shape[1];
+    output_shape_.resize(N, dim);
+    for (int i = 0; i < N; i++) {
+      for (int d = 0; d < dim; d++) {
+        output_shape_.tensor_shape_span(i)[d] = *shape_tensor(i, d);
+      }
+    }
+  } else {
+    int N = shape.num_samples();
+    int sample_dim;
+    for (int i = 0; i < N; i++) {
+      int current_sample_dim = shape.tensor_shape_span(i)[0];
+      if (i) {
+        DALI_ENFORCE(current_sample_dim == sample_dim,
+          "Reshape: all samples must have the same number of dimensions");
+      } else {
+        sample_dim = current_sample_dim;
+        output_shape_.resize(N, sample_dim);
+      }
+
+      for (int d = 0; d < sample_dim; d++) {
+        output_shape_.tensor_shape_span(i)[d] = shape.tensor_data(i)[d];
+      }
+    }
+  }
+}
+
+template <typename Backend>
+template <typename TensorListLike>
+void Reshape<Backend>::ShapeFromInput(const TensorListLike &tl) {
+  TYPE_SWITCH(tl.type().id(), type2id, type, (int32_t),
```

Comment:
don't you want to support other integral types? If not, please rephrase the DALI_FAIL below to say that int32 is expected